File: client/src/types/tutor-session-context.ts
export interface TutorSessionContext {
  schema_version: 'home-tutor.tutor-session-context.v1'
  page: 'session_review'
  session_id: string
  question_id: string
  navigation: {
    question_index: number
    total_questions: number
    timeline_ms?: number
  }
  package: {
    prompt_text: string
    options?: Array<{ key: string; text: string }>
    final_answer: string
    question_type?: string
    verdict: 'correct' | 'wrong' | 'unknown'
    process_summary?: string
  }
  tutor: {
    summary: string
    error_category?: string
    explanation_paragraphs: string[]
    analysis_status: string
  }
  a2ui: {
    surface_id: string
    data_model: Record<string, unknown>
    recent_actions: Array<{ id: string; ts: number }>
  }
  transcript: Array<{
    role: 'user' | 'assistant'
    content: string
    source: 'asr' | 'text' | 'action'
    ts: number
  }>
  vision_hints?: Array<{
    region_id: string
    reason: string
    layout_hash: string
  }>
}

export type TranscriptEntry = TutorSessionContext['transcript'][number]

File: server/src/home_tutor/models/tutor_session_context.py
"""Tutor session context for AG-UI chat rounds."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class NavigationContext(BaseModel):
    """Session navigation slice sent to the tutor chat agent."""

    question_index: int
    total_questions: int
    timeline_ms: int | None = None


class PackageContext(BaseModel):
    """Question package slice for chat context."""

    prompt_text: str
    options: list[dict[str, str]] | None = None
    final_answer: str
    question_type: str | None = None
    verdict: Literal["correct", "wrong", "unknown"]
    process_summary: str | None = None


class TutorContextSlice(BaseModel):
    """Existing tutor analysis slice."""

    summary: str
    error_category: str | None = None
    explanation_paragraphs: list[str]
    analysis_status: str


class A2UIContextSlice(BaseModel):
    """A2UI surface state."""

    surface_id: str
    data_model: dict[str, Any] = Field(default_factory=dict)
    recent_actions: list[dict[str, Any]] = Field(default_factory=list)


class TranscriptEntry(BaseModel):
    """One chat turn."""

    role: Literal["user", "assistant"]
    content: str
    source: Literal["asr", "text", "action"]
    ts: int


class VisionHint(BaseModel):
    """Metadata for an optional vision attachment."""

    region_id: str
    reason: str
    layout_hash: str


class TutorSessionContext(BaseModel):
    """Full structured context for one tutor chat round."""

    schema_version: Literal["home-tutor.tutor-session-context.v1"] = (
        "home-tutor.tutor-session-context.v1"
    )
    page: Literal["session_review"]
    session_id: str
    question_id: str
    navigation: NavigationContext
    package: PackageContext
    tutor: TutorContextSlice
    a2ui: A2UIContextSlice
    transcript: list[TranscriptEntry] = Field(default_factory=list)
    vision_hints: list[VisionHint] | None = None
